PerimeterXModule/Internals/CredentialsIntelligence/PxLoginData.cs

```diff
@@ -0,0 +1,226 @@
+        }
+
+        public ExtractedCredentials ExtractCredentials(ExtractorObject extractionDetails, PxContext pxContext, HttpRequest request)
+        {
+            string userFieldName = extractionDetails.UserFieldName;
+            string passwordFieldName = extractionDetails.PassFieldName;
+
+            Dictionary<string, string> headers = pxContext.GetHeadersAsDictionary();
+
+            if (userFieldName == null || passwordFieldName == null)
+            {
+                return null;
+            }
+
+            if (extractionDetails.SentThrough == "header")
+            {
+                return ExtractFromHeader(userFieldName, passwordFieldName, headers);
+            } else if (extractionDetails.SentThrough == "query-param")
+            {
+                return new ExtractedCredentials(
+                    request.QueryString[userFieldName].Replace(" ", "+"),
+                    request.QueryString[passwordFieldName].Replace(" ", "+")
+                );
+            } else if (extractionDetails.SentThrough == "body")
+            {
+                return ExtractFromBodyAsync(userFieldName, passwordFieldName, headers, request).Result;
+            }
+
+            return null;
+        }
+
+        public static ExtractedCredentials ExtractFromHeader(string userFieldName, string passwordFieldName, Dictionary<string, string> headers)
+        {
+            bool isUsernameHeaderExist = headers.TryGetValue(userFieldName, out string userName);
+            bool isPasswordHeaderExist = headers.TryGetValue(passwordFieldName, out string password);
+
+            if (!isUsernameHeaderExist && !isPasswordHeaderExist) { return null; }
+
+            return new ExtractedCredentials(userName, password);
+        }
+
+        public async Task<ExtractedCredentials> ExtractFromBodyAsync(string userFieldName, string passwordFieldName, Dictionary<string, string> headers, HttpRequest request)
+        {
+            bool isContentTypeHeaderExist = headers.TryGetValue("Content-Type", out string contentType);
+
+            HttpRequest httpRequest = request;
+
+            string body = await ReadRequestBodyAsync(httpRequest);
+
+            if (!isContentTypeHeaderExist)
+            {
+                return null;
+            } else if (contentType.Contains("application/json"))
+            {
+                return ExtractCredentialsFromJson(body, userFieldName, passwordFieldName);
+            } else if (contentType.Contains("x-www-form-urlencoded"))
+            {
+                return ReadValueFromUrlEncoded(body, userFieldName, passwordFieldName);
+            } else if (contentType.Contains("form-data"))
+            {
+                return ExtarctValueFromMultipart(body, contentType, userFieldName, passwordFieldName);
+            }
+            
+            return null;
+        }
+
+        public static async Task<string> ReadRequestBodyAsync(HttpRequest request)
+        {
+            HttpRequest httpRequest = request;
```

Comment:
Why? Does this make a copy or something?

Reply:
Yes, I created a copy of the request because it's not possible to read the body more than once.